scheduler/src/cook/mesos/util.clj

```diff
@@ -338,21 +344,72 @@
+                         (< submit-ms end-ms))))
+
+;; get-jobs-by-user-and-state-and-submit is a bit opaque
+;; because it is reaching into datomic internals. Here is a quick explanation.
+;; seek-datoms provides a pointer into the raw datomic indices
+;; that we can then seek through. We set the pointer to look
+;; through the avet index, with attribute :job/state, seek to
+;; state and then seek to the entity id that *would* have been
+;; created at expanded start. This works because the submission
+;; time (which we sort on) is correlated with the entitiy id.
+;; We then seek through the list of all jobs in that state,
+;; then filtering to the target user (which is cached).
+;;
+;; This function is O(# jobs in that state in the time range)
+(defn get-jobs-by-user-and-state-and-submit
+  "Returns all jobs for a particular user in the specified state
+   and timeframe. Returns lazy output."
+  [db ^String user ^Date start ^Date end state-keyword]
+  (let [;; Expand the time range so that clock skew between cook
+        ;; and datomic doesn't cause us to miss jobs
+        ;; 1 hour was picked because a skew larger than that would be
+        ;; suspicious
+        expanded-start (Date. (- (.getTime start)
+                                 (-> 1 t/hours t/in-millis)))
+        expanded-end (Date. (+ (.getTime end)
+                               (-> 1 t/hours t/in-millis)))
+        entid-start (d/entid-at db :db.part/user expanded-start)
+        entid-end (d/entid-at db :db.part/user expanded-end)
+        job-state-entid (d/entid db :job/state)
+        state-entid (d/entid db state-keyword)
+        start-ms (.getTime start)
+        end-ms (.getTime end)]
+    (->> (d/seek-datoms db :avet :job/state state-entid entid-start)
+         (take-while #(and (< (:e %) entid-end)
+                           (= (:a %) job-state-entid)
+                           (= (:v %) state-entid)))
+         (map :e)
+         (map (partial d/entity db))
+         (filter #(= (job-ent->user %) user))
+         (filter #(job-submit-time-between % start-ms end-ms)))))
+
 ;; get-completed-jobs-by-user is a bit opaque because it is
 ;; reaching into datomic internals. Here is a quick explanation.
-;; seek-datoms provides a pointer into the raw datomic indices 
+;; seek-datoms provides a pointer into the raw datomic indices
 ;; that we can then seek through. We set the pointer to look
-;; through the avet index, with attribute :job/user, seek to 
+;; through the avet index, with attribute :job/user, seek to
 ;; user and then seek to the entity id that *would* have been
-;; created at expanded start. 
+;; created at expanded start.
 ;; This works because the submission time and job/user field
 ;; are set at the same time, in "real" time. This means that
 ;; jobs submitted after `start` will have been created after
 ;; expanded start
+;; This differs from get-active-jobs-by-user-and-state as it
+;; is also looking up based on task state.
```

Comment:
Can you add a comment that this is O(number of jobs the user has submitted) to mirror the one above?

Reply:
Done